pkg/antctl/raw/upgrade/apistorage/command.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return err
+	}
+
+	crdObjectsToUpgrade := getCRDsToUpgrade(cmd, opts.k8sClient, crdNamesToUpgrade)
+	for _, crd := range crdObjectsToUpgrade {
+		if err = upgradeCRDObjects(cmd, opts.k8sClient, crd); err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Failed to upgrade CRD %q: %v.\n", crd.Name, err)
+		}
+		if err = patchCRDStoredVersions(opts.k8sClient, crd); err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Failed to patch CRD %q status.storedVersion: %v.\n", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// getAntreaCRDNames gets names of all Antrea CRDs.
+func getAntreaCRDNames(k8sClient client.Client) (sets.Set[string], error) {
+	crdNames := sets.New[string]()
+	crdList := &apiextv1.CustomResourceDefinitionList{}
+	if err := k8sClient.List(context.TODO(), crdList); err != nil {
+		return nil, fmt.Errorf("failed to get CRD list: %v", err)
+	}
+	for _, crd := range crdList.Items {
+		if strings.HasSuffix(crd.Name, "crd.antrea.io") {
+			crdNames.Insert(crd.Name)
+		}
+	}
+
+	return crdNames, nil
+}
+
+// getCRDNamesToUpgrade gets names of Antrea CRDs to upgrade.
+func getCRDNamesToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNamesToUpgrade sets.Set[string]) (sets.Set[string], error) {
+	antreaCRDNames, err := getAntreaCRDNames(k8sClient)
+	if err != nil {
+		return nil, err
+	}
+	// If the user-provided name list of CRDs to upgrade is empty, upgrade all Antrea CRDs.
+	if crdNamesToUpgrade.Len() == 0 {
+		crdNamesToUpgrade = antreaCRDNames
+	} else {
+		// If the user-provided name list of CRDs to upgrade is not empty, and it contains CRDs without suffix
+		// "crd.antrea.io", skip these CRDs.
+		for name := range crdNamesToUpgrade.Difference(antreaCRDNames) {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip CRD %q which is not created by Antrea.\n", name)
+		}
+		// Only upgrade the CRDs with suffix "crd.antrea.io".
+		crdNamesToUpgrade = crdNamesToUpgrade.Intersection(antreaCRDNames)
+	}
+	return crdNamesToUpgrade, nil
+}
+
+// getCRDsToUpgrade gets a list of Antrea CRDs to upgrade.
+func getCRDsToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) []*apiextv1.CustomResourceDefinition {
+	var crdsToUpgrade []*apiextv1.CustomResourceDefinition
+	for crdName := range crdNameListToUpgrade {
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Failed to get CRD %q: %v.\n", crdName, err)
+			continue
+		}
+		// Skip the CRD that has only one version.
+		if len(crd.Spec.Versions) < 2 {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip upgrading CRD %q since it only has one version.\n", crdName)
+			continue
+		}
+		crdsToUpgrade = append(crdsToUpgrade, crd)
+	}
+	return crdsToUpgrade
+}
+
+// upgradeCRDObjects upgrades the existing objects of a CRD.
+func upgradeCRDObjects(cmd *cobra.Command, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
```

Comment:
It seems the function doesn't require cmd but the writer only. It should make tests easier.

Reply:
Updated as you said.